File: backend/app/receipt/schema.py
from typing import List
from pydantic import BaseModel
from datetime import datetime


receipt_schema = {
    "type": "object",
    "properties": {
        "restaurant": {"type": "string", "description": "Name of the restaurant"},
        "address": {"type": "string", "description": "Address of the restaurant"},
        "opened": {
            "type": "string",
            "description": "Date and time the order was opened",
        },
        "order_number": {"type": "string", "description": "Unique order number"},
        "order_type": {
            "type": "string",
            "description": "Type of the order (e.g., dine-in, takeout)",
        },
        "table": {"type": "string", "description": "Table number or identifier"},
        "server": {"type": "string", "description": "Name or ID of the server"},
        "items": {
            "type": "array",
            "description": "List of items ordered",
            "items": {
                "type": "object",
                "properties": {
                    "name": {
                        "type": "string",
                        "description": "Name of the ordered item",
                    },
                    "price": {
                        "type": "number",
                        "description": "Price of the ordered item",
                    },
                    "quantity": {
                        "type": "integer",
                        "description": "Quantity of the ordered item",
                    },
                },
                "required": ["name", "price", "quantity"],
                "additionalProperties": False,
            },
        },
        "subtotal": {"type": "number", "description": "Subtotal before tax"},
        "sales_tax": {"type": "number", "description": "Sales tax amount"},
        "total": {"type": "number", "description": "Total amount of the order"},
        "payment": {
            "type": "object",
            "description": "Payment information",
            "properties": {
                "method": {
                    "type": "string",
                    "description": "Payment method (e.g., cash, credit card)",
                },
                "amount_paid": {"type": "number", "description": "Total amount paid"},
                "tip": {"type": "number", "description": "Tip amount given"},
            },
            "required": ["method", "amount_paid", "tip"],
            "additionalProperties": False,
        },
        "copy": {
            "type": "string",
            "description": "Receipt copy type (e.g., customer, merchant)",
        },
        "other_fees": {
            "type": "array",
            "description": "List of additional fees applied to the order",
            "items": {
                "type": "object",
                "properties": {
                    "name": {
                        "type": "string",
                        "description": "Name of the additional fee",
                    },
                    "price": {
                        "type": "number",
                        "description": "Price of the additional fee",
                    },
                },
                "required": ["name", "price"],
                "additionalProperties": False,
            },
        },
    },
    "required": [
        "restaurant",
        "address",
        "order_number",
        "order_type",
        "table",
        "server",
        "items",
        "subtotal",
        "sales_tax",
        "total",
        "payment",
        "copy",
        "other_fees",
        "opened",
    ],
    "additionalProperties": False,
}


class OrderItem(BaseModel):
    name: str
    price: float
    quantity: int


class PaymentInfo(BaseModel):
    method: str
    amount_paid: float
    tip: float


class OtherFee(BaseModel):
    name: str
    price: float


class Receipt(BaseModel):
    restaurant: str
    address: str
    opened: datetime | None
    order_number: str
    order_type: str
    table: str
    server: str
    items: List[OrderItem]
    subtotal: float
    sales_tax: float
    total: float
    payment: PaymentInfo
    copy: str
    other_fees: List[OtherFee]

File: backend/internal/receipt/schema.go
package receipt

import "time"

type Receipt struct {
	Restaurant  string         `json:"restaurant" jsonschema_description:"Name of the restaurant"`
	Address     string         `json:"address" jsonschema_description:"Address of the restaurant"`
	Opened      string         `json:"opened" jsonschema_description:"Date and time the order was opened"`
	OrderNumber string         `json:"order_number" jsonschema_description:"Unique order number"`
	OrderType   string         `json:"order_type" jsonschema_description:"Type of the order (e.g., dine-in, takeout)"`
	Table       string         `json:"table" jsonschema_description:"Table number or identifier"`
	Server      string         `json:"server" jsonschema_description:"Name or ID of the server"`
	Items       []OrderItem    `json:"items" jsonschema_description:"List of items ordered"`
	Subtotal    float64        `json:"subtotal" jsonschema_description:"Subtotal before tax"`
	SalesTax    float64        `json:"sales_tax" jsonschema_description:"Sales tax amount"`
	Total       float64        `json:"total" jsonschema_description:"Total amount of the order"`
	Payment     PaymentDetails `json:"payment" jsonschema_description:"Payment information"`
	Copy        string         `json:"copy" jsonschema_description:"Receipt copy type (e.g., customer, merchant)"`
	OtherFees   []OtherFee     `json:"other_fees" jsonschema_description:"List of additional fees applied to the order"`
}

type OrderItem struct {
	Name     string  `json:"name" jsonschema_description:"Name of the ordered item"`
	Price    float64 `json:"price" jsonschema_description:"Price of the ordered item"`
	Quantity int     `json:"quantity" jsonschema_description:"Quantity of the ordered item"`
}

type PaymentDetails struct {
	Method     string  `json:"method" jsonschema_description:"Payment method (e.g., cash, credit card)"`
	AmountPaid float64 `json:"amount_paid" jsonschema_description:"Total amount paid"`
	Tip        float64 `json:"tip" jsonschema_description:"Tip amount given"`
}

type OtherFee struct {
	Name  string  `json:"name" jsonschema_description:"Name of the additional fee"`
	Price float64 `json:"price" jsonschema_description:"Price of the additional fee"`
}

type ParsedReceipt struct {
	Receipt
	Opened time.Time
}
